### PR TITLE
Add ChunkCache: support get vector from storage

### PR DESCRIPTION
**Pull request by bigsheeper**
issue: https://github.com/milvus-io/milvus/issues/26141

/kind enhancement

### CODE REVIEW

**Comment by bigsheeper** [comment]
see also: https://github.com/milvus-io/milvus/issues/21296

File: internal/core/src/mmap/Column.h [hunk 1]
```diff
@@ -81,6 +81,23 @@ class ColumnBase {
                                         0));
     }
 
+    // mmap mode ctor
+    ColumnBase(const File& file,
+               size_t size,
+               size_t num_rows,
```
Comment (yah01) [inline internal/core/src/mmap/Column.h:87]: miss to update `num_rows_`?

File: internal/core/src/mmap/Column.h [hunk 2]
```diff
@@ -81,6 +81,23 @@ class ColumnBase {
                                         0));
     }
 
+    // mmap mode ctor
+    ColumnBase(const File& file,
```
Comment (yah01) [inline internal/core/src/mmap/Column.h:85]: why we need this ctor?

File: internal/core/src/segcore/SegmentSealedImpl.cpp [hunk 3]
```diff
@@ -590,12 +600,34 @@ SegmentSealedImpl::vector_search(SearchInfo& search_info,
     }
 }
 
+std::tuple<std::string, int64_t>
+SegmentSealedImpl::get_field_data_path(FieldId field_id, int64_t offset) const {
```
Comment (yah01) [inline internal/core/src/segcore/SegmentSealedImpl.cpp:604]: To keep style consistent, let's rename this following camelCase

File: internal/core/src/segcore/SegmentSealedImpl.cpp [hunk 4]
```diff
@@ -452,6 +453,15 @@ SegmentSealedImpl::LoadDeletedRecord(const LoadDeletedRecordInfo& info) {
...
 }
 
+void
+SegmentSealedImpl::AddFieldDataInfo(const LoadFieldDataInfo& field_data_info) {
+    // deep copy
+    for (const auto& iter : field_data_info.field_infos) {
+        field_data_info_.field_infos.emplace(iter.first, iter.second);
+    }
+    field_data_info_.mmap_dir_path = field_data_info.mmap_dir_path;
+}
```
Comment (yah01) [inline internal/core/src/segcore/SegmentSealedImpl.cpp:463]: simply `field_data_info_ = field_data_info` should work well

File: internal/core/src/segcore/SegmentSealedImpl.cpp [hunk 5]
```diff
@@ -590,12 +600,34 @@ SegmentSealedImpl::vector_search(SearchInfo& search_info,
...
+std::tuple<std::string, int64_t>
+SegmentSealedImpl::get_field_data_path(FieldId field_id, int64_t offset) const {
+    auto offset_in_binlog = offset;
+    auto data_path = std::string();
+    AssertInfo(field_data_info_.field_infos.find(field_id.get()) !=
+                   field_data_info_.field_infos.end(),
+               fmt::format("cannot find binlog file for field: {}, seg: {}",
+                           field_id.get(),
+                           id_));
+    auto field_info = field_data_info_.field_infos.at(field_id.get());
```
Comment (yah01) [inline internal/core/src/segcore/SegmentSealedImpl.cpp:612]: the following code could be faster as only seeking data once:
```cpp
auto it = field_data_info_.find(field_id.get());
Assert(it != field_data_info_.end(), ...);
auto& field_info = it->second);
```

**Review by yah01 — CHANGES_REQUESTED** [review]

File: internal/core/src/mmap/Column.h [hunk 6]
```diff
@@ -81,6 +81,23 @@ class ColumnBase {
...
+               size_t size,
+               size_t num_rows,
+               const DataType& data_type) {
+        padding_ = data_type == DataType::JSON ? simdjson::SIMDJSON_PADDING : 0;
+
+        size_ = size;
+        cap_size_ = size;
```
Comment (czs007) [inline internal/core/src/mmap/Column.h:92]: should calculate type_size

File: internal/core/src/mmap/Column.h [hunk 7]
```diff
@@ -115,6 +132,11 @@ class ColumnBase {
         return num_rows_;
     };
 
+    const size_t
+    ByteSize() const {
```
Comment (czs007) [inline internal/core/src/mmap/Column.h:136]: Size() ？ @yah01

File: internal/querynodev2/server.go [hunk 8]
```diff
@@ -221,8 +221,24 @@ func (node *QueryNode) InitSegcore() error {
...
 
+	err := initcore.InitRemoteChunkManager(paramtable.Get())
+	if err != nil {
+		return err
+	}
+
+	mmapDirPath := paramtable.Get().QueryNodeCfg.MmapDirPath.GetValue()
+	if len(mmapDirPath) == 0 {
+		// Fallback to local path, TODO: or read from remote each time get vector?
+		mmapDirPath = paramtable.Get().LocalStorageCfg.Path.GetValue()
```
Comment (czs007) [inline internal/querynodev2/server.go:232]: add a sub dir mmap?

File: internal/querynodev2/segments/segment.go [hunk 9]
```diff
@@ -744,6 +744,56 @@ func (s *LocalSegment) LoadFieldData(fieldID int64, rowCount int64, field *datap
...
+		}
+
+		for _, binlog := range field.Binlogs {
+			err = loadFieldDataInfo.appendLoadFieldDataPath(fieldID, binlog)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	loadFieldDataInfo.appendMMapDirPath(paramtable.Get().QueryNodeCfg.MmapDirPath.GetValue())
```
Comment (czs007) [inline internal/querynodev2/segments/segment.go:782]: should check: wrong mmap if user not config QueryNodeCfg.MmapDirPath?

File: internal/core/src/storage/Util.cpp [hunk 10]
```diff
@@ -433,6 +453,40 @@ GetObjectData(ChunkManager* remote_chunk_manager,
     return datas;
 }
 
+std::map<std::string, int64_t>
+PutFieldData(ChunkManager* remote_chunk_manager,
```
Comment (czs007) [inline internal/core/src/storage/Util.cpp:457]: Move to unit test directory

File: internal/core/src/mmap/Column.h (hunk 7 above)
Comment (yah01) [reply internal/core/src/mmap/Column.h:136]: it's not the same, `Size()` returns the length of column (like `std::vector::size()`), and `ByteSize()` returns the size of the column in bytes

File: internal/core/src/storage/ChunkCache.cpp [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+    auto path = std::filesystem::path(path_prefix_) / filepath;
+    ColumnTable::const_accessor ca;
+    if (!columns_.find(ca, path)) {
+        return;
+    }
+    auto column = ca->second;
+    auto ok =
+        madvise(reinterpret_cast<void*>(const_cast<char*>(column->Data())),
+                column->ByteSize(),
+                MADV_SEQUENTIAL);
```
Comment (yah01) [inline internal/core/src/storage/ChunkCache.cpp:59]: we'd better to test the performance with this, as `MADV_SEQUENTIAL` aggressively reads ahead and may free the mapping soon. The access pattern of columns is sequential right, but it's random in reduce stage. `MADV_WILLNEED` and `MADV_RANDOM` should be also options.

**Review by yah01 — COMMENTED** [review]
/lgtm

File: internal/core/src/storage/ChunkCache.cpp [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+    }
+    ca.release();
+
+    auto object_data =
+        GetObjectData(cm_.get(), std::vector<std::string>{filepath});
+    AssertInfo(object_data.size() == 1, "GetObjectData failed");
+    auto field_data = object_data[0];
+
+    auto column = Mmap(path, field_data);
+    columns_.emplace(path, column);
```
Comment (zhagnlu) [inline internal/core/src/storage/ChunkCache.cpp:38]: thread safe is not valid

File: internal/core/src/segcore/segment_c.cpp [hunk 13]
```diff
@@ -366,3 +366,21 @@ DropSealedSegmentIndex(CSegmentInterface c_segment, int64_t field_id) {
...
+
+CStatus
+AddFieldDataInfo(CSegmentInterface c_segment,
+                 CLoadFieldDataInfo c_load_field_data_info) {
+    try {
+        auto segment_interface =
+            reinterpret_cast<milvus::segcore::SegmentInterface*>(c_segment);
+        auto segment =
+            dynamic_cast<milvus::segcore::SegmentSealed*>(segment_interface);
+        AssertInfo(segment != nullptr, "segment conversion failed");
```
[the remark this answers is not in the document]
Comment (bigsheeper) [reply internal/core/src/segcore/segment_c.cpp:378]: yes, it is only for sealed; Renamed.

**Comment by bigsheeper** [comment]
rerun ut

**Comment by bigsheeper** [comment]
rerun ut

**Comment by czs007** [comment]
@bigsheeper please remember to pick to 2.3.0

**Comment by bigsheeper** [comment]
> @bigsheeper please remember to pick to 2.3.0

https://github.com/milvus-io/milvus/pull/27062

**Pull request merged**